pylsp/python_lsp.py

```diff
@@ -375,11 +382,72 @@ def hover(self, doc_uri, position):
         # Since we're debounced, the document may no longer be open
         workspace = self._match_uri_to_workspace(doc_uri)
-        if doc_uri in workspace.documents:
-            workspace.publish_diagnostics(
-                doc_uri,
-                flatten(self._hook('pylsp_lint', doc_uri, is_saved=is_saved))
-            )
+        document_object = workspace.documents.get(doc_uri, None)
+        if isinstance(document_object, Document):
+            self._lint_text_document(doc_uri, workspace, is_saved=is_saved)
+        elif isinstance(document_object, Notebook):
+            self._lint_notebook_document(document_object, workspace)
+
+    def _lint_text_document(self, doc_uri, workspace, is_saved):
+        workspace.publish_diagnostics(
+            doc_uri,
+            flatten(self._hook('pylsp_lint', doc_uri, is_saved=is_saved))
+        )
+
+    def _lint_notebook_document(self, notebook_document, workspace):
+        """
+        Lint a notebook document.
+
+        This is a bit more complicated than linting a text document, because we need to
+        send the entire notebook document to the pylsp_lint hook, but we need to send
+        the diagnostics back to the client on a per-cell basis.
+        """
+
+        # First, we create a temp TextDocument that represents the whole notebook
+        # contents. We'll use this to send to the pylsp_lint hook.
+        random_uri = str(uuid.uuid4())
+
+        # cell_list helps us map the diagnostics back to the correct cell later.
+        cell_list: List[Dict[str, Any]] = []
+
+        offset = 0
+        total_source = ""
+        for cell in notebook_document.cells:
+            cell_uri = cell['document']
+            cell_document = workspace.get_cell_document(cell_uri)
+
+            num_lines = len(cell_document.lines)
+
+            data = {
+                'uri': cell_uri,
+                'line_start': offset + 1,
+                'line_end': offset + num_lines,
+                'source': cell_document.source
+            }
+
+            cell_list.append(data)
+            total_source = total_source + "\n" + cell_document.source
+
+            offset += num_lines
+
+        workspace.put_document(random_uri, total_source)
+        document_diagnostics = flatten(self._hook('pylsp_lint', random_uri, is_saved=True))
+
+        # Now we need to map the diagnostics back to the correct cell and
+        # publish them.
+        for cell in cell_list:
+            cell_diagnostics = []
+            for diagnostic in document_diagnostics:
+                if diagnostic['range']['start']['line'] > cell['line_end']:
+                    break
+                diagnostic['range']['start']['line'] = diagnostic['range']['start']['line'] - cell['line_start'] + 1
+                diagnostic['range']['end']['line'] = diagnostic['range']['end']['line'] - cell['line_start'] + 1
+                cell_diagnostics.append(diagnostic)
+                document_diagnostics.pop(0)
```

Comment:
Is this needed?

```suggestion
```

Reply:
This was intended, but actually dangerous in a for loop, which is what I realized later. I wanted to keep the diagnostics splitting "efficient", which is why I removed already assigned diagnostics. I opted for a computationally more complex version now, which "just works". Let's optimze later.